File: java/year2021/month2/no124/BinaryTreeMaximumPathSum.java
package year2021.month2.no124;

public class BinaryTreeMaximumPathSum {
    private static int max = Integer.MIN_VALUE;

    public static void main(String[] args) {
        TreeNode root1 = new TreeNode(1, new TreeNode(2), new TreeNode(3));
        TreeNode root2 = new TreeNode(-10,
                new TreeNode(9),
                new TreeNode(20, new TreeNode(15), new TreeNode(7)));
        TreeNode root3 = new TreeNode(-10, new TreeNode(2), new TreeNode(3));
        TreeNode root4 = new TreeNode(2, new TreeNode(-1), null);
        System.out.println(maxPathSum(root1));
        System.out.println(maxPathSum(root2));
        System.out.println(maxPathSum(root3));
        System.out.println(maxPathSum(root4));
    }

    public static int maxPathSum(TreeNode root) {
        max = Integer.MIN_VALUE;
        dfs(root);
        return max;
    }

    private static int dfs(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int leftVal = dfs(root.left);
        int rightVal = dfs(root.right);
        if (leftVal <= 0 && rightVal <= 0) {
            max = Math.max(max, root.val);
            return root.val;
        }
        int nodeMaxVal = root.val + Math.max(leftVal, rightVal);
        int pathMaxVal = root.val + leftVal + rightVal;
        max = Math.max(max, Math.max(nodeMaxVal, pathMaxVal));
        return nodeMaxVal;
    }
}

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

/*
 * 路径 被定义为一条从树中任意节点出发，沿父节点-子节点连接，达到任意节点的序列。该路径 至少包含一个 节点，且不一定经过根节点。

路径和 是路径中各节点值的总和。

给你一个二叉树的根节点 root ，返回其 最大路径和 。

 

示例 1：


输入：root = [1,2,3]
输出：6
解释：最优路径是 2 -> 1 -> 3 ，路径和为 2 + 1 + 3 = 6
示例 2：


输入：root = [-10,9,20,null,null,15,7]
输出：42
解释：最优路径是 15 -> 20 -> 7 ，路径和为 15 + 20 + 7 = 42
 

提示：

树中节点数目范围是 [1, 3 * 104]
-1000 <= Node.val <= 1000

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/binary-tree-maximum-path-sum
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
* */

File: go/year2025/month3/no124/BinaryTreeMaximumPathSum.go
package main

func main() {
	root1 := &TreeNode{
		Val: 1,
		Left: &TreeNode{
			Val: 2,
		},
		Right: &TreeNode{
			Val: 3,
		},
	}
	println(maxPathSum(root1)) //6
	root2 := &TreeNode{
		Val: -10,
		Left: &TreeNode{
			Val: 9,
		},
		Right: &TreeNode{
			Val: 20,
			Left: &TreeNode{
				Val: 15,
			},
			Right: &TreeNode{
				Val: 7,
			},
		},
	}
	println(maxPathSum(root2)) //42
	root3 := &TreeNode{Val: -3}
	println(maxPathSum(root3)) //-3
	root4 := &TreeNode{Val: 1,
		Left: &TreeNode{Val: -2,
			Left:  &TreeNode{Val: 1},
			Right: &TreeNode{Val: 3},
		},
		Right: &TreeNode{Val: -3,
			Right: &TreeNode{Val: -1},
		},
	}
	println(maxPathSum(root4)) //3
}

// dfs, times: O(n), space: O(n)
func maxPathSum(root *TreeNode) int {
	cMax, _ := dfs(root)
	return cMax
}

// 返回值：
//   - finalMax 不包含根节点的最大路径和 - 已经是最大路径，不可再与 parent 连接
//   - rootMax 包含根节点的最大路径和 - 可能与 parent 连接
func dfs(root *TreeNode) (finalMax int, rootMax int) {
	if root == nil {
		return -9999, -9999
	}
	lf, lr := dfs(root.Left)
	rf, rr := dfs(root.Right)
	// 三种情况： 1. 左子树的路径最大 ; 2. 右子树的路径最大; 3. 以 root 为根节点的路径最大
	finalMax = max(lf, rf, root.Val, root.Val+lr, root.Val+rr, root.Val+lr+rr)
	rootMax = max(root.Val, root.Val+lr, root.Val+rr) // 一定要带上 root 才能再给 parent 连接
	return
}

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/*
二叉树中的 路径 被定义为一条节点序列，序列中每对相邻节点之间都存在一条边。
同一个节点在一条路径序列中 至多出现一次 。该路径 至少包含一个 节点，且不一定经过根节点。

路径和 是路径中各节点值的总和。
给你一个二叉树的根节点 root ，返回其 最大路径和 。

示例 1：
输入：root = [1,2,3]
输出：6
解释：最优路径是 2 -> 1 -> 3 ，路径和为 2 + 1 + 3 = 6

示例 2：
输入：root = [-10,9,20,null,null,15,7]
输出：42
解释：最优路径是 15 -> 20 -> 7 ，路径和为 15 + 20 + 7 = 42


提示：
树中节点数目范围是 [1, 3 * 104]
-1000 <= Node.val <= 1000
*/
